panels/user-accounts/um-app-permissions.c

```diff
@@ -0,0 +1,787 @@
+          !g_app_info_should_show (app) ||
+          app_name[0] == '\0' ||
+          /* Endless' link apps have the "eos-link" prefix, and should be ignored too */
+          g_str_has_prefix (g_app_info_get_id (app), "eos-link"))
+        {
+          continue;
+        }
+
+      g_debug ("Processing app '%s' (%s)", g_app_info_get_id (app), g_app_info_get_executable (app));
+
+      g_list_store_insert_sorted (self->apps,
+                                  app,
+                                  compare_app_info_cb,
+                                  self);
+    }
+
+  g_list_free_full (apps, g_object_unref);
+}
+
+static GsContentRatingSystem
+get_content_rating_system (ActUser *user)
+{
+  const gchar *user_language;
+  gchar *str;
+
+  user_language = act_user_get_language (user);
+
+  /* Remove the encoding suffix */
+  str = g_strstr_len (user_language, -1, ".UTF-8");
+  if (str != NULL)
+    *str = '\0';
+  str = g_strstr_len (user_language, -1, ".utf8");
+  if (str != NULL)
+    *str = '\0';
+
+  return gs_utils_content_rating_system_from_locale (user_language);
+}
+
+static void
+schedule_update_blacklisted_apps (UmAppPermissions *self)
+{
+  if (self->blacklist_apps_source_id > 0)
+    return;
+
+  self->blacklist_apps_source_id = g_timeout_add_seconds (1, blacklist_apps_cb, self);
+}
+
+static void
+update_app_filter (UmAppPermissions *self)
+{
+  g_autoptr(GError) error = NULL;
+
+  g_clear_pointer (&self->filter, epc_app_filter_unref);
+  self->filter = epc_get_app_filter (NULL,
+                                     act_user_get_uid (self->user),
+                                     FALSE,
+                                     self->cancellable,
+                                     &error);
+
+  if (error)
+    {
+      g_warning ("Error retrieving app filter for user '%s': %s",
+                  act_user_get_user_name (self->user),
+                  error->message);
+      return;
+    }
+
+  g_debug ("Retrieved new app filter for user '%s'", act_user_get_user_name (self->user));
+}
+
+static void
+update_categories_from_language (UmAppPermissions *self)
+{
+  GsContentRatingSystem rating_system;
+  const gchar * const * entries;
+  const gchar *rating_system_str;
+  const guint *ages;
+  guint i;
+
+  rating_system = get_content_rating_system (self->user);
+  rating_system_str = gs_content_rating_system_to_str (rating_system);
+
+  g_debug ("Using system %s", rating_system_str);
+
+  entries = gs_utils_content_rating_get_values (rating_system);
+  ages = gs_utils_content_rating_get_ages (rating_system);
+
+  /* Fill in the age menu */
+  g_menu_remove_all (self->age_menu);
+
+  for (i = 0; entries[i] != NULL; i++)
+    {
+      g_autofree gchar *action = g_strdup_printf ("permissions.set-age(uint32 %u)", ages[i]);
+
+      g_menu_append (self->age_menu, entries[i], action);
+    }
+}
+
+/* Returns an human-readable but untranslated string, not suitable
+ * to be shown in any UI */
+static inline const gchar*
```

Comment:
Aside: I’d question the use of `inline` here. The compiler will already inline this if it makes sense to do so.